### PR TITLE
WIP: generate Warthog model files from templates

### PR DESCRIPTION
**Pull request by dzhelezov**
This is a WIP PR (built on top of https://github.com/Joystream/joystream/pull/641) for https://github.com/Joystream/joystream/issues/673
All the key features are already available.

TODOs:
- [x]  Replace hardcoded paths and namings with calls to warthog's Config
- [x]  Refactor: unify FTS and Model files rendering, extract interfaces, separate rendering and path resolution
- [x] Test: Cover with unittests key entity relations and types

### CODE REVIEW

File: query-node/substrate-query-framework/cli/src/templates/entities/model.ts.mst [hunk 1]
```diff
@@ -0,0 +1,59 @@
+import {
+  BaseModel,
+  {{#has.bool}}BooleanField,{{/has.bool}}
+  {{#has.date}}DateField,{{/has.date}}
+  {{#has.float}}FloatField,{{/has.float}}
+  {{#has.int}}IntField,{{/has.int}}
+  {{#has.numeric}}NumericField,{{/has.numeric}}
+  {{#has.json}}JSONField,{{/has.json}}
+  {{#has.bytes}}JSONField,{{/has.bytes}}
```
Comment (metmirr) [inline query-node/substrate-query-framework/cli/src/templates/entities/model.ts.mst]: Should be `BytesField`

**Comment by metmirr** [comment]
Very impressive work!

Except for relationships (`mto`, `otm`...) it works well with https://github.com/Joystream/joystream/issues/525

There are naming and typescript types problems with related fields. For example with this schema

```graphql
type Author @entity {
  posts: [Post!]
}

type Post @entity {
  title: String
  author: Author!
}
```

column definition for `posts` field will be following:

```typescript
@OneToMany(() => ---, (posts: Posts) => posts.posts)
posts?: Posts[];
```

actually it should be:

```typescript
@OneToMany(() => Post, (post: Post) => post.author)
posts?: Post[];
```

File: query-node/substrate-query-framework/cli/src/templates/entities/model.ts.mst (hunk 1 above)
Comment (dzhelezov) [reply query-node/substrate-query-framework/cli/src/templates/entities/model.ts.mst]: Fixed + added some unit tests

**Comment by dzhelezov** [comment]
> Very impressive work!
> 
> Except for relationships (`mto`, `otm`...) it works well with #525
> 
> There are naming and typescript types problems with related fields. For example with this schema
> 
> ```graphql
> type Author @entity {
>   posts: [Post!]
> }
> 
> type Post @entity {
>   title: String
>   author: Author!
> }
> ```
> 
> column definition for `posts` field will be following:
> 
> ```ts
> @OneToMany(() => ---, (posts: Posts) => posts.posts)
> posts?: Posts[];
> ```
> 
> actually it should be:
> 
> ```ts
> @OneToMany(() => Post, (post: Post) => post.author)
> posts?: Post[];
> ```

Fixed and added tests. Now there's a convenient method `tests/model/fromSchemaString` for testing various combinations with inline schemas, so we can easily extend the tests as more bugs are spotted.

**Review by metmirr — APPROVED** [review]

**Review by bedeho — APPROVED** [review]
Epic work, really nicely done. I am just glancing through here, can't provide any quality assurance.

**Pull request merged**